fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexBuildState.java

```diff
@@ -0,0 +1,174 @@
+ * </p>
+ * <p>
+ * Note that index build state (especially records scanned) serves an informational purpose only as the number of
+ * records scanned is not necessarily 100% of records in total when the index build completes for several reasons:
+ * </p>
+ * <ul>
+ *     <li> If the {@link OnlineIndexer} has been configured to disable tracking progress (by setting
+ *     {@link OnlineIndexer.Builder#setTrackProgress(boolean)} to {@code false}), then {@link #getRecordsScanned()} will
+ *     not reflect the actual number of records scanned. </li>
+ *     <li> Records added or deleted during the index build progress are not included in records scanned but are in
+ *     records in total. </li>
+ *     <li> Records in total uses the number of all records in the store. In some cases, the real number of records
+ *     that need to be scanned is less than this.</li>
+ * </ul>
+ * @see #loadIndexBuildStateAsync(FDBRecordStoreBase, Index)
+ */
+@API(API.Status.UNSTABLE)
+public class IndexBuildState {
+    @Nonnull
+    private final IndexState indexState;
+    @Nullable
+    private final Long recordsScanned;
+    @Nullable
+    private final Long recordsInTotal;
+
+    /**
+     * Load the build progress ({@link IndexBuildState}) of the given index in the given record store asynchronously.
+     * @param store the record store containing the index
+     * @param index the index needed to be checked
+     * @return a future that completes to the index build state
+     */
+    @Nonnull
+    public static CompletableFuture<IndexBuildState> loadIndexBuildStateAsync(FDBRecordStoreBase<?> store, Index index) {
+        IndexState indexState = store.getUntypedRecordStore().getIndexState(index);
+        if (indexState != IndexState.WRITE_ONLY) {
+            return CompletableFuture.completedFuture(new IndexBuildState(indexState));
+        }
+        CompletableFuture<Long> recordsInTotalFuture;
+        try {
+            recordsInTotalFuture = store.getSnapshotRecordCount();
+        } catch (AggregateFunctionNotSupportedException ex) {
+            // getSnapshotRecordCount failed, very likely it is because there is no suitable COUNT type index
+            // defined.
+            recordsInTotalFuture = CompletableFuture.completedFuture(null);
+        }
+        return loadRecordsScannedAsync(store, index).thenCombine(recordsInTotalFuture,
+                (scannedRecords, recordsInTotal) -> new IndexBuildState(indexState, scannedRecords, recordsInTotal));
+    }
+
+    /**
+     * Load the number of records successfully scanned and processed during the online index build process
+     * asynchronously.
+     * <p>
+     * If the {@link OnlineIndexer} has been configured to disable tracking progress (by setting
+     * {@link OnlineIndexer.Builder#setTrackProgress(boolean)} to {@code false}), then the number returned will not
+     * reflect the actual number of records scanned.
```

Comment:
In this and `getRecordsScanned`, when the tracking feature is disabled, the phrase `the number returned will not reflect that actual number of records scanned` is a little ambiguous, particularly given that the `Long` returned can also be `null`.

Reply:
Changed to "then the number returned will not include the records scanned by that {@link OnlineIndexer}"